filebeat/input/tcp/client.go

```diff
@@ -0,0 +1,118 @@
+package tcp
+
+import (
+	"bufio"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Client is a remote client.
+type Client struct {
+	conn          net.Conn
+	forwarder     *harvester.Forwarder
+	done          chan struct{}
+	metadata      common.MapStr
+	splitFunc     bufio.SplitFunc
+	maxReadBuffer int64
+	timeout       time.Duration
+}
+
+// NewClient returns a new client instance for the remote connection.
+func NewClient(
+	conn net.Conn,
+	forwarder *harvester.Forwarder,
+	splitFunc bufio.SplitFunc,
+	maxReadBuffer int64,
+	timeout time.Duration,
+) *Client {
+	client := &Client{
+		conn:          conn,
+		forwarder:     forwarder,
+		done:          make(chan struct{}),
+		splitFunc:     splitFunc,
+		maxReadBuffer: maxReadBuffer,
+		timeout:       timeout,
+	}
+	client.cacheMetadata()
+	return client
+}
+
+// Handle is reading message from the specified TCP socket.
+func (c *Client) Handle() error {
+	defer c.conn.Close()
+
+	r := NewMeteredReader(NewDeadlineReader(c.conn, c.timeout), c.maxReadBuffer)
+	buf := bufio.NewReader(r)
+	scanner := bufio.NewScanner(buf)
+	scanner.Split(c.splitFunc)
+
+	for scanner.Scan() {
+		if scanner.Err() != nil {
+			// we are forcing a close on the socket, lets ignore any error that could happen.
+			select {
+			case <-c.done:
+				break
+			default:
+			}
+			// This is a user defined limit and we should notify the user.
+			if IsMaxReadBufferErr(scanner.Err()) {
+				logp.Err("tcp client error: %s", scanner.Err())
+			}
+			return errors.Wrap(scanner.Err(), "tcp client error")
+		}
+		r.Reset()
+		// TODO(ph), should we keep the byte array instead of allocating a string?
+		c.forwarder.Send(c.createEvent(scanner.Text()))
+	}
+	return nil
+}
+
+// Close stops reading from the socket and close the connection.
+func (c *Client) Close() {
+	close(c.done)
+	c.conn.Close()
+}
+
+func (c *Client) createEvent(rawString string) *util.Data {
+	data := util.NewData()
+	data.Event = beat.Event{
+		Timestamp: time.Now(),
+		Meta:      c.metadata,
+		Fields: common.MapStr{
+			"message": rawString,
+		},
+	}
+	return data
+}
+
+// GetRemoveHosts take the IP address of the client and try to resolve the name, if it fails we
+// fallback to the IP, IP can resolve to multiple hostname.
+func (c *Client) getRemoteHosts() []string {
+	ip := c.conn.RemoteAddr().String()
+	idx := strings.Index(ip, ":")
+	if idx == -1 {
+		return []string{ip}
+	}
+	ip = ip[0:idx]
+	hosts, err := net.LookupAddr(ip)
+	if err != nil {
+		hosts = []string{ip}
+	}
+	return hosts
+}
+
+func (c *Client) cacheMetadata() {
+	c.metadata = common.MapStr{
+		"hostnames":  c.getRemoteHosts(),
```

Comment:
Should we report all hosts? Should we exclude some?

Reply:
I am not sure what is the best behavior here:

1. taking the first one could be wrong.
2. taking a subset could also be wrong.
3. They are set in the metadata field for now, so users are feed to use them or not.

Reply:
Any chance you could provide some example json output docs in the comment of the PR? I think that makes it easier to understand how the end event will look like.

Reply:
Also wonder if we want to opt-out from resolving the hostname.

Reply:
Right it could be costly or slow.